src/main/scala/com/fulcrumgenomics/fastq/FastqSource.scala

```diff
@@ -0,0 +1,113 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.fulcrumgenomics.fastq
+
+import java.io._
+
+import com.fulcrumgenomics.util.Io
+import dagr.commons.CommonsDef.{PathToFastq, yieldAndThen}
+
+/**
+  * Provides factory methods for creating a `FastqSource` from multiple types.
+  */
+object FastqSource {
+  /** Creates a new fastq source from a Reader. */
+  def apply(reader: Reader): FastqSource = new FastqSource(reader)
+
+  /** Creates a new fastq source from a File. */
+  def apply(file: File): FastqSource = apply(path=file.toPath)
+
+  /** Creates a new fastq source from a Path. */
+  def apply(path: PathToFastq): FastqSource = apply(Io.toReader(path))
+}
+
+
+/**
+  * Reads fastq records from any text based source via a reader. Ensures that lines come in
+  * the expected groupings of four lines with the correct headers, and that bases and qualities
+  * are of the same length.  The underlying reader is closed automatically when EOF is reached.
+  */
+class FastqSource private(in: Reader) extends Iterator[FastqRecord] with Closeable {
+  private val reader : BufferedReader = in match {
+    case in : BufferedReader => in
+    case in : Reader         => new BufferedReader(in)
+  }
+
+  private var nextRecord: Option[FastqRecord] = None
+  queueNextRecord()
+
+  /** True if calling `next()` will yield another record, false otherwise. */
+  override def hasNext: Boolean = this.nextRecord.isDefined
+
+  /** Returns the next record if available, or throws an exception if none is available. */
+  override def next(): FastqRecord =  yieldAndThen(nextRecord.get) { queueNextRecord() }
+
+  /** Short hand to throw an illegal state exception. */
+  private def illegal(msg: String): Nothing = throw new IllegalStateException(msg)
+
+  /** Sets the current record to None, and then attempts to read the next record from the input. */
+  private def queueNextRecord() : Unit = {
+    this.nextRecord = None
+    val header  = reader.readLine()
+    val seq     = reader.readLine()
+    val qHeader = reader.readLine()
+    val quals   = reader.readLine()
+
+    if (header == null) {
+      this.reader.close()
+    }
+    else {
+      // Assert that we read four lines or no lines
+      if (header != null && (seq == null || qHeader == null || quals == null)) {
+        illegal(s"Fastq source terminated mid-record at ${header}")
+      }
+
+      if (!header.startsWith("@"))  illegal(s"Fastq sequence header must start with @: ${header}")
+      if (!qHeader.startsWith("+")) illegal(s"Fastq quality header must start with @: ${qHeader}")
+      if (seq.length != quals.length) illegal(s"Sequence and qualities not same length for record: ${header}")
+
+      // Destructure the header line into name, read number and comment
+      val (fullName, comment) = header.indexWhere(_.isWhitespace) match {
```

Comment:
I did think about that, but decided that the above would work and should be significantly more efficient that relying on a regex.